internal/pkg/service/appproxy/appconfig/appconfig.go

```diff
@@ -0,0 +1,106 @@
+package appconfig
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/keboola/go-client/pkg/client"
+	"github.com/keboola/go-client/pkg/request"
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+type Loader struct {
+	logger log.Logger
+	clock  clock.Clock
+	sender request.Sender
+	cache  map[string]cacheItem
+}
+
+type cacheItem struct {
+	config    AppProxyConfig
+	eTag      string
+	expiresAt time.Time
+}
+
+func NewLoader(logger log.Logger, clock clock.Clock, baseURL string) *Loader {
+	return &Loader{
+		logger: logger,
+		clock:  clock,
+		sender: client.New().WithBaseURL(baseURL),
+		cache:  make(map[string]cacheItem),
+	}
+}
+
+const staleCacheFallbackDuration = time.Hour
+
+func (l *Loader) LoadConfig(ctx context.Context, appID string) (AppProxyConfig, error) {
+	var config *AppProxyConfig
+	var err error
+	now := l.clock.Now()
+
+	if item, ok := l.cache[appID]; ok {
+		// Return config from cache if still valid
+		if now.Before(item.expiresAt) {
+			return item.config, nil
+		}
+
+		// API request with cached ETag
+		config, err = GetAppProxyConfig(l.sender, appID, item.eTag).Send(ctx)
+		if err != nil {
+			return l.handleError(ctx, appID, now, err, &item)
+		}
+
+		// Update expiration and use the cached config if ETag is still the same
+		if config.eTag == item.eTag {
+			l.cache[appID] = cacheItem{
+				config:    item.config,
+				eTag:      item.eTag,
+				expiresAt: now.Add(config.maxAge),
```

Comment:
What if we set a maximum to be sure, e.g. `1 hour` ?

Reply:
It can be simple implemented using generic types, but it is not present in the std lib:
https://github.com/golang/go/issues/50019